### PR TITLE
Handle stability of struct fields

### PR DESCRIPTION
**Pull request by huonw**
We were recording stability attributes applied to fields in the
compiler, and even annotating it in the libs, but the compiler didn't
actually do the checks to give errors/warnings in user crates.

Details in the commit messages.

### CODE REVIEW

**Comment by huonw** [comment]
r? @brson, @aturon, @alexcrichton 

NB. I've had to add 3 stability attributes.

**Comment by alexcrichton** [comment]
This looks good to me, nice work! Could this also be expanded to cover fields mentioned in patterns as well for structs? Other than that I think it covers everything.

File: src/librustc/middle/stability.rs [hunk 1]
```diff
@@ -356,6 +361,76 @@ pub fn check_expr(tcx: &ty::ctxt, e: &ast::Expr,
...
+            span = field.span;
+            match ty::expr_ty_adjusted(tcx, base_e).sty {
+                ty::ty_struct(did, _) => {
+                    ty::lookup_struct_fields(tcx, did)
+                        .iter()
+                        .find(|f| f.name == field.node.name)
+                        .unwrap_or_else(|| {
+                            tcx.sess.span_bug(field.span,
+                                              "stability::check_expr: unknown named field access")
+                        })
```
Comment (alexcrichton) [inline src/librustc/middle/stability.rs]: I wonder if this sort of lookup would be candidate for a good helper in `middle::ty`

Comment (huonw) [inline src/librustc/middle/stability.rs]: I think how we handle structs in the compiler will be changing with https://github.com/rust-lang/rust/pull/22564, so I'll leave it for now.

Comment (alexcrichton) [inline src/librustc/middle/stability.rs]: Ah ok, this can wait then. r=me with patterns handled as well

**Comment by alexcrichton** [comment]
(cancelled build as we discovered some interesting tidbits on IRC)

**Comment by alexcrichton** [comment]
@bors: retry

nevermind

**Comment by Manishearth** [comment]
(Manual cancel due to rollup)

**Pull request merged**